File: samples/java/DeliveryDispatch/Server/CourierSession/src/main/java/systems/zlink/samples/deliverydispatch/server/couriersession/sessions/CourierSession.java
package systems.zlink.samples.deliverydispatch.server.couriersession.sessions;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ActorRefSnapshot;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionActor;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class CourierSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers;
    private final ZLinkActorManager actors;

    public CourierSession(
            ZLinkSessionContext context,
            ZLinkSessionPacketDispatcher<ZLinkSessionContext> handlers,
            ZLinkActorManager actors) {
        this.context = context;
        this.handlers = handlers;
        this.actors = actors;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        CompletableFuture<?>[] notifications =
                context.actors().bound().stream()
                        .map(actor -> actor.notifyDisconnected().toCompletableFuture())
                        .toArray(CompletableFuture[]::new);
        return CompletableFuture.allOf(notifications);
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        if ("BindCourierSessionReq".equals(dispatch.packetName())) {
            return handleBindCourierSessionReq(dispatch, payload);
        }
        return handlers.tryHandle(context, dispatch, payload)
                .thenCompose(
                        handled -> {
                            if (handled) {
                                return CompletableFuture.completedFuture(null);
                            }
                            Messages.CourierDecisionMsg decision =
                                    payload.decode(Messages.CourierDecisionMsg.class);
                            ZLinkSessionActor actor =
                                    context.actors()
                                            .find(decision.courierId())
                                            .orElseThrow(
                                                    () ->
                                                            new IllegalStateException(
                                                                    "Courier actor is not bound: "
                                                                            + decision
                                                                                    .courierId()));
                            return actor.relay(payload).thenApply(ignored -> null);
                        });
    }

    // --8<-- [start:doc-dd-session-bind]
    private CompletionStage<Void> handleBindCourierSessionReq(
            ZLinkSessionDispatchContext dispatch, ZLinkMessage payload) {
        Messages.BindCourierSessionReq request =
                payload.decode(Messages.BindCourierSessionReq.class);
        return findOrEnsureActor(request.courierId())
                .thenCompose(
                        actorRef -> {
                            ZLinkSessionActor bound =
                                    context.actors().find(actorRef.actorId()).orElse(null);
                            boolean requiresBinding = bound == null;
                            CompletionStage<ZLinkSessionActor> actorStage =
                                    bound == null
                                            ? context.actors().bind(actorRef)
                                            : CompletableFuture.completedFuture(bound);
                            ActorRefSnapshot snapshot = ActorRefSnapshot.from(actorRef);
                            Messages.BindCourierSessionReq relayed =
                                    new Messages.BindCourierSessionReq(
                                            request.courierId(), snapshot, context.sessionId());
                            return actorStage.thenCompose(
                                    actor -> {
                                        if (requiresBinding) {
                                            System.out.println(
                                                    "deliverydispatch-courier bound courier="
                                                            + request.courierId());
                                        }
                                        return actor.relay(dispatch, ZLinkMessage.of(relayed))
                                                .thenApply(ignored -> null);
                                    });
                        });
    }

    // --8<-- [end:doc-dd-session-bind]

    private CompletionStage<ActorRef> findOrEnsureActor(String courierId) {
        return actors.getOrCreate(courierId, SampleNames.CourierActorType)
                .request(new Messages.EnsureCourierActorReq(courierId))
                .submit()
                .thenApply(
                        result -> {
                            if (result instanceof ZLinkActorCreateResult.Existing existing) {
                                return existing.actor();
                            }
                            if (result instanceof ZLinkActorCreateResult.Created created) {
                                return created.actor();
                            }
                            throw new IllegalStateException("Courier Actor creation was rejected.");
                        });
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSession/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/couriersession/sessions/CourierSession.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.couriersession.sessions

import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ActorRefSnapshot
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkSessionPacketDispatcher
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.BindCourierSessionReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.BindCourierSessionRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.CourierDecisionMsg
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.EnsureCourierActorReq

class CourierSession(
    private val sessionContext: ZLinkSessionContext,
    private val handlers: ZLinkSessionPacketDispatcher<ZLinkSessionContext>,
    private val actors: ZLinkActorManager,
) : ZLinkSuspendingSession() {
    override fun context(): ZLinkSessionContext = sessionContext

    override suspend fun onConnectedSuspending() {}

    override suspend fun onDisconnectedSuspending() {
        for (actor in sessionContext.actors().bound()) actor.notifyDisconnected().await()
    }

    override suspend fun onErrorSuspending(error: ZLinkStreamError) {}

    override suspend fun onDispatchSuspending(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        if (dispatch.packetName() == "BindCourierSessionReq") {
            handleBindCourierSessionReq(dispatch, payload)
            return
        }
        val handled = handlers.tryHandle(sessionContext, dispatch, payload).await()
        if (handled) {
            return
        }
        val decision = payload.decode(CourierDecisionMsg::class.java)
        val actor =
            sessionContext.actors().find(decision.courierId).orElseThrow {
                IllegalStateException("Courier actor is not bound: ${decision.courierId}")
            }
        actor.relay(dispatch, payload).await()
    }

    // --8<-- [start:doc-dd-session-bind]
    private suspend fun handleBindCourierSessionReq(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        val request = payload.decode(BindCourierSessionReq::class.java)
        val actorRef = findOrEnsureActor(request.courierId)
        val actor =
            sessionContext.actors().find(actorRef.actorId).orElse(null)
                ?: sessionContext.actors().bind(actorRef).await()
        val snapshot = ActorRefSnapshot.from(actorRef)
        actor
            .relay(
                dispatch,
                ZLinkMessage.of(
                    BindCourierSessionReq(
                        courierId = request.courierId,
                        actor = snapshot,
                        sessionRoute = sessionContext.sessionId(),
                    )
                ),
            )
            .await()
        sessionContext
            .client()
            .reply(BindCourierSessionRes(request.courierId, snapshot, sessionContext.sessionId()))
            .submit()
        println("deliverydispatch-courier bound courier=${request.courierId}")
    }

    // --8<-- [end:doc-dd-session-bind]

    private suspend fun findOrEnsureActor(courierId: String): ActorRef =
        when (
            val result =
                actors
                    .kotlin()
                    .getOrCreate(courierId, SampleNames.CourierActorType)
                    .request(EnsureCourierActorReq(courierId))
                    .await()
        ) {
            is ZLinkActorCreateResult.Existing -> result.actor
            is ZLinkActorCreateResult.Created -> result.actor
            is ZLinkActorCreateResult.Rejected ->
                throw IllegalStateException("Courier Actor creation was rejected.")
        }
}
